File: bench/fit_phylo_nb2.jl
# fit_phylo_nb2.jl — Julia side of the NB2 phylo benchmark.
#
# Run from repo root after `Rscript bench/R/gen_phylo_nb2.R`:
#   julia --project=bench bench/fit_phylo_nb2.jl

import Pkg
Pkg.activate(dirname(@__DIR__))
using DRM
using LinearAlgebra
using DelimitedFiles, Printf, Statistics

BLAS.set_num_threads(1)

const FIXDIR = joinpath(@__DIR__, "fixtures", "phylo_nb2")
const OUTDIR = joinpath(@__DIR__, "results", "phylo_nb2")
mkpath(OUTDIR)

const CELLS = [
    (id = "phylo_p100", p = 100, n = 500, reps = 5),
    (id = "phylo_p500", p = 500, n = 1500, reps = 3),
    (id = "phylo_p1000", p = 1000, n = 2000, reps = 3),
    (id = "phylo_p2000", p = 2000, n = 4000, reps = 2),
]

function read_fixture(id)
    raw = readdlm(joinpath(FIXDIR, "$id.csv"), ',', String; header = true)[1]
    y = Float64.(parse.(Int, raw[:, 1]))
    x = parse.(Float64, raw[:, 2])
    species = raw[:, 3]
    tree = read(joinpath(FIXDIR, "$id.nwk"), String)
    return y, x, species, tree
end

json_str(x::AbstractString) = "\"" * replace(x, "\\" => "\\\\", "\"" => "\\\"") * "\""
json_val(x::Bool) = x ? "true" : "false"
json_val(x::Integer) = string(x)
json_val(x::Real) = isfinite(x) ? @sprintf("%.17g", x) : "null"
json_val(x::AbstractString) = json_str(x)
json_val(x::AbstractVector) = "[" * join(json_val.(x), ",") * "]"
json_val(::Nothing) = "null"
function json_obj(pairs)
    "{" * join([json_str(String(k)) * ":" * json_val(v) for (k, v) in pairs], ",") * "}"
end

function fit_cell(cell)
    y, x, species, tree = read_fixture(cell.id)
    drm(bf(@formula(y ~ x + phylo(1 | species)), @formula(sigma ~ 1)),
        NegBinomial2(); data = (; y, x, species), tree = tree, g_tol = 1e-6, se = false)
end

function run_cell(cell)
    fit_cell(cell)
    times = Float64[]
    fits = Any[]
    for _ in 1:cell.reps
        fit = nothing
        t = @elapsed begin
            fit = fit_cell(cell)
        end
        push!(times, t)
        push!(fits, fit)
    end
    fit = fits[end]
    rs = re_sd(fit)
    theta = exp(coef(fit, :sigma)[1])
    row = [
        :cell_id => cell.id,
        :engine => "julia_sparse_laplace",
        :p => cell.p,
        :n => cell.n,
        :time_s => mean(times),
        :time_s_med => median(times),
        :times_all => times,
        :logLik => loglik(fit),
        :converged => fit.converged,
        :beta_mu => collect(coef(fit, :mu)),
        :theta_nb2 => theta,
        :sd_phylo => get(rs, :species, NaN),
        :julia_threads => Threads.nthreads(),
        :blas_threads => BLAS.get_num_threads(),
    ]
    @printf("[Julia %s] p=%d n=%d med=%.4fs logLik=%.3f beta=(%.3f, %.3f) theta=%.3f sd=%.3f\n",
            cell.id, cell.p, cell.n, row[6][2], row[8][2],
            row[10][2][1], row[10][2][2], row[11][2], row[12][2])
    return row
end

results = [run_cell(cell) for cell in CELLS]
open(joinpath(OUTDIR, "julia_phylo_nb2.json"), "w") do io
    println(io, "[")
    for (i, row) in enumerate(results)
        comma = i == length(results) ? "" : ","
        println(io, "  ", json_obj(row), comma)
    end
    println(io, "]")
end

File: bench/R/fit_phylo_nb2.R
## fit_phylo_nb2.R — drmTMB side of the NB2 phylo benchmark.
##
## Run from repo root after fixtures exist:
##   Rscript bench/R/fit_phylo_nb2.R

suppressPackageStartupMessages({
  local_drmTMB <- "/Users/z3437171/Dropbox/Github Local/drmTMB"
  if (dir.exists(local_drmTMB) && requireNamespace("devtools", quietly = TRUE)) {
    devtools::load_all(local_drmTMB, quiet = TRUE)
  } else if (requireNamespace("drmTMB", quietly = TRUE)) {
    library(drmTMB)
  } else {
    stop("drmTMB not installed and devtools not available to load_all().")
  }
  library(ape)
  library(jsonlite)
})

`%||%` <- function(a, b) if (is.null(a)) b else a

here <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  script_path <- if (length(file_arg)) {
    normalizePath(sub("^--file=", "", file_arg[[1]]))
  } else {
    normalizePath(sys.frames()[[1]]$ofile %||% ".")
  }
  dirname(dirname(script_path))
}

bench_root <- tryCatch(here(), error = function(e) file.path(getwd(), "bench"))
fixtures_dir <- file.path(bench_root, "fixtures", "phylo_nb2")
results_dir <- file.path(bench_root, "results", "phylo_nb2")
dir.create(results_dir, showWarnings = FALSE, recursive = TRUE)

cells <- list(
  list(id = "phylo_p100", p = 100, n = 500, reps = 5),
  list(id = "phylo_p500", p = 500, n = 1500, reps = 3),
  list(id = "phylo_p1000", p = 1000, n = 2000, reps = 3),
  list(id = "phylo_p2000", p = 2000, n = 4000, reps = 2)
)

coef_dpar <- function(fit, dpar) {
  cf <- coef(fit)
  if (is.list(cf) && !is.null(cf[[dpar]])) return(unname(as.numeric(cf[[dpar]])))
  numeric()
}

phylo_sd <- function(fit) {
  sp <- fit$sdpars$mu
  if (is.null(sp) || !length(sp)) return(NA_real_)
  nm <- names(sp)
  hit <- grep("phylo|species", nm, ignore.case = TRUE)
  if (!length(hit)) hit <- seq_along(sp)[[1]]
  unname(as.numeric(sp[[hit[[1]]]]))
}

fit_one <- function(cell) {
  df <- read.csv(file.path(fixtures_dir, paste0(cell$id, ".csv")))
  tr <- ape::read.tree(file.path(fixtures_dir, paste0(cell$id, ".nwk")))
  df$species <- factor(df$species, levels = tr$tip.label)
  form <- drmTMB::bf(y ~ x + phylo(1 | species, tree = tr), sigma ~ 1)
  ctrl <- drmTMB::drm_control(se = FALSE)

  warm <- tryCatch(
    drmTMB::drmTMB(form, data = df, family = nbinom2(), control = ctrl),
    error = function(e) structure(list(error = conditionMessage(e)), class = "drm_phylo_err")
  )
  if (inherits(warm, "drm_phylo_err")) {
    return(list(
      cell_id = cell$id, engine = "r_drmTMB", p = cell$p, n = cell$n,
      time_s = NA_real_, time_s_med = NA_real_, times_all = rep(NA_real_, cell$reps),
      logLik = NA_real_, converged = FALSE, beta_mu = numeric(),
      beta_sigma = numeric(), sigma_nb2 = NA_real_, theta_nb2 = NA_real_,
      sd_phylo = NA_real_, note = paste("warm-up failed:", warm$error)
    ))
  }

  times <- rep(NA_real_, cell$reps)
  fits <- vector("list", cell$reps)
  for (k in seq_len(cell$reps)) {
    t <- system.time({
      fits[[k]] <- drmTMB::drmTMB(form, data = df, family = nbinom2(), control = ctrl)
    })
    times[[k]] <- as.numeric(t[["elapsed"]])
  }
  fit <- fits[[length(fits)]]
  beta <- coef_dpar(fit, "mu")
  beta_sigma <- coef_dpar(fit, "sigma")
  sigma_nb2 <- if (length(beta_sigma)) exp(beta_sigma[[1L]]) else NA_real_
  theta_nb2 <- if (is.finite(sigma_nb2) && sigma_nb2 > 0) 1 / sigma_nb2^2 else NA_real_
  res <- list(
    cell_id = cell$id,
    engine = "r_drmTMB",
    p = cell$p,
    n = cell$n,
    time_s = mean(times),
    time_s_med = stats::median(times),
    times_all = times,
    logLik = as.numeric(logLik(fit)),
    converged = isTRUE(fit$opt$convergence == 0L),
    beta_mu = beta,
    beta_sigma = beta_sigma,
    sigma_nb2 = sigma_nb2,
    theta_nb2 = theta_nb2,
    sd_phylo = phylo_sd(fit)
  )
  cat(sprintf("[R %s] p=%d n=%d med=%.4fs logLik=%.3f beta=(%.3f, %.3f) theta=%.3f sd=%.3f\n",
              cell$id, cell$p, cell$n, res$time_s_med, res$logLik,
              beta[[1]], beta[[2]], res$theta_nb2, res$sd_phylo))
  res
}

results <- lapply(cells, fit_one)
jsonlite::write_json(results, file.path(results_dir, "r_phylo_nb2.json"),
                     auto_unbox = TRUE, pretty = TRUE, digits = NA)
